### PR TITLE
I4: 2:4 structured sparsity path (mask ramp + cuSPARSELt-format export)

### PR DESCRIPTION
**Pull request by gmboquet**
## Summary

Implements roadmap item **I4** (2:4 structured sparsity path): a schedulable training-time mask ramp plus a real cuSPARSELt-style compressed export/decompress round trip.

**Stacked on `sigma-weighted-projections` (PR #144 / G2)** — this PR's diff (`mixle/models/sparsity_2_4.py`, `mixle/tests/sparsity_2_4_test.py`) is only the two new files; it should be reviewed/merged after #144 lands, and only shows a clean diff once that branch is merged.

- `mixle/models/sparsity_2_4.py`:
  - `TwoFourSparsityRamp(start_step, end_step, target_density=0.5, schedule=None)` — ramps the *fraction of a weight matrix's rows* under the hard 2:4 constraint from 0% to 100% between `start_step`/`end_step`. It does **not** reimplement 2:4 masking: at every ramp step it calls G2's `sigma_weighted_block_sparse(w, sigma, "2:4")` directly on the currently-constrained rows (identity `Sigma` by default, i.e. plain magnitude-based value selection; a real propagated G1 covariance can be passed in too). The step→fraction map is a swappable `schedule` callable — a deliberate seam a future roadmap-H3 "structure-edit schedule" controller could plug into (not built here).
  - `export_2_4_compressed` / `decompress` — a real cuSPARSELt-shaped compressed-matrix format: 2 surviving values per group of 4 plus a packed 2-bit-per-position index (2 groups packed per byte). Exact round trip, verified on real 2:4-masked matrices including all-zero/partial-zero edge-case groups.
  - `cusparselt_status()` — queries this torch build/environment honestly (`torch.cuda.is_available()`, `torch.backends.cusparselt`, `torch.sparse.SparseSemiStructuredTensor`) rather than assuming capability.
- `mixle/tests/sparsity_2_4_test.py` — the acceptance tests (see below).

## Environment check (this matters for the "measured speedup" claim)

```
torch 2.12.1, cuda_available=False, cusparselt_is_available=False
torch.sparse.to_sparse_semi_structured(...) on CPU -> RuntimeError (CUDA-only, regardless of cuSPARSELt)
```

No CUDA device exists in this environment, so **no real cuSPARSELt-accelerated 2:4 GEMM was run or could be run here.** The tests/report below are explicit about what's *measured* vs. *theoretical*.

## Measured / reported numbers

**Loss parity** (small transformer, `d_model=16, n_layer=2, n_head=4`, Markov-chain synthetic next-token task, 120 training steps, identical init/data for both runs):
- entropy floor (irreducible loss): **2.5736 nats**
- dense-trained held-out loss: **3.1483 nats**
- 2:4-ramp-trained held-out loss: **3.1760 nats**
- gap: **0.0276 nats** (well inside the stated 0.35-nat tolerance; both models also land within 2x of the entropy floor)

**Compression ratio** (real matrix, 64x256, packed indices):
- dense: 131072 bytes, compressed: 67584 bytes (65536 values + 2048 indices) → **1.939x**

**Inference speedup — honestly split:**
- *Theoretical* (vendor claim on real sparse-tensor-core kernels, **not measured here**): 2:4 halves the surviving multiply-adds → **2.00x FLOP-reduction bound**; measured compressed-byte-size vs dense → **1.88x memory-traffic reduction bound**.
- *Measured* (real wall-clock on this CPU, **explicitly not a cuSPARSELt number**): dense 1024x1024 matmul = 0.218 ms vs CPU unstructured-CSR sparse matmul = 10.61 ms (**0.021x — CSR is ~46x *slower*, as expected: CPU sparse BLAS overhead dominates at 50% density; unstructured CSR is also not 2:4-aware**). Reported honestly rather than omitted or dressed up.
- `torch.sparse.to_sparse_semi_structured` confirmed to raise on this CPU-only build (no CUDA tensor available), consistent with `cusparselt_status()`.

## Test plan

- [x] `mixle/tests/sparsity_2_4_test.py` — 15 tests (mask-ramp correctness incl. G2-delegation check, compress/decompress round trip + measured compression ratio, loss parity, honest speedup evidence) — all pass
- [x] `mixle/tests/sigma_weighted_projection_test.py` (G2 regression) — 8 tests, all pass
- [x] `ruff check --fix` / `ruff format` on both new files

### CODE REVIEW

**Comment by gmboquet** [comment]
Added a real GPU cuSPARSELt measurement, run on a rented NVIDIA RTX A4000 (Ampere, cc 8.6), to sit alongside this PR's honest CPU-only CSR result.

Real, measured (not theoretical) speedups from `torch.sparse.to_sparse_semi_structured`:
- 1024x1024 @ batch=64: **0.51x** (slower than dense -- indexing overhead dominates at this size)
- 4096x4096 @ batch=64: **1.65x**
- 4096x4096 @ batch=256: **1.20x**
- 8192x8192 @ batch=128: **2.09x** (essentially the 2.0x theoretical FLOP-reduction bound)

New test: `InferenceSpeedupEvidenceTest.test_measured_gpu_wall_clock_dense_vs_real_cusparselt_semi_structured_matmul`, CUDA-gated so it skips cleanly on the CPU-only dev machine and only runs where a real cuSPARSELt-capable GPU is present.

**Comment by gmboquet** [comment]
Relanded as #209 (merged).
